Added fuzzer-to-job configurability on /jobs page

iamarshsingh opened the pull request:
Hi, This PR is to add Fuzzer Selection option to /jobs page. The relationship defined between the jobs and fuzzer is many-to-many and the ground truth for the relationships is considered to be Fuzzer.jobs with automatic updation to JobFuzzer as well. This PR will fixes #596. 

Screenshot of the location of the menu:
![Screenshot 2020-07-13 at 4 47 09 PM](https://user-images.githubusercontent.com/30122295/87298909-affa4680-c528-11ea-9720-f7aecc974c68.png)


Screenshot of the drop-down:
![Screenshot 2020-07-13 at 4 44 39 PM](https://user-images.githubusercontent.com/30122295/87298708-411ced80-c528-11ea-9d08-db6f9ed7f72a.png)

Please have a look and let me know if there are any changes required!

iamarshsingh commented:
/gcbrun

iamarshsingh commented:
/gcbrun

inferno-chromium commented on src/appengine/private/components/jobs/multi-listbox.html:
```diff
@@ -0,0 +1,81 @@
+<!--
```
Shouldn't this go in components/common directory ? Also, might be nice to use it in other pages where we have multi list selection ?

iamarshsingh replied:
Actually multi-listbox is just a wrapper around paper-listbox with IronFormElementBehavior. This was only required in the jobs page as we were using iron-form. Other pages where multi-list selection is required uses ajax-requests which doesn't require the wrapper.

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -123,6 +140,39 @@ def post(self):
+    # Retriving old associated fuzzers.
+    query = data_types.Fuzzer.query()
+    query = query.filter(data_types.Fuzzer.jobs == name)
+    old_fuzzers = ndb_utils.get_all_from_query(query)
```
nit: don't use get_all_from_query in new code. It's not longer needed and will be removed at some point.

iamarshsingh replied:
done

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -67,9 +68,23 @@ def get_results(this):
   items, total_pages, total_items, has_more = query.fetch_page(
       page=page, page_size=PAGE_SIZE, projection=None, more_limit=MORE_LIMIT)
 
+  dic_items = []
```
+1. Would name it something like "format_jobs" or something similar though.

iamarshsingh replied:
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -123,6 +140,39 @@ def post(self):
           'Job name can only contain letters, numbers, dashes and underscores.',
           400)
 
+    # Retriving old associated fuzzers.
```
+1. We can go with the same approach, but put this in its own function.

iamarshsingh replied:
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -123,6 +140,39 @@ def post(self):
+    for fuzzer in old_fuzzers:
+      old_mappings[fuzzer.name] = fuzzer
+
+    # Adding new associated fuzzers.
```
We want to retain Fuzzer.jobs as the source of truth from with FuzzerJobs are generated, so went with this using update_mappings_for_fuzzer. Adding another place that modifies FuzzerJob adds more maintenance burden with little gain.

iamarshsingh replied:
I have kept the same functionality as @oliverchang has mentioned. The complexity increase won't be much in the current builds as there are not a lot of jobs, but if in future the number of jobs increases a lot, we can maybe shift to changing FuzzerJob directly. If we should change it now, please let me know!

iamarshsingh commented:
/gcbrun

iamarshsingh commented:
Hi @inferno-chromium and @oliverchang thank you for the review. I have addressed all the changes and added unittests for both the new functionalities. Please have a look!

iamarshsingh commented on src/appengine/handlers/jobs.py:
```diff
@@ -67,9 +68,23 @@ def get_results(this):
   items, total_pages, total_items, has_more = query.fetch_page(
       page=page, page_size=PAGE_SIZE, projection=None, more_limit=MORE_LIMIT)
 
+  dic_items = []
```
done 👍

iamarshsingh commented on src/appengine/handlers/jobs.py:
```diff
@@ -67,9 +68,23 @@ def get_results(this):
+    query = data_types.Fuzzer.query()
+    query = query.filter(data_types.Fuzzer.jobs == item.name)
+    fuzzers = ndb_utils.get_all_from_query(query)
+    dic_item['fuzzers'] = []
```
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -53,6 +53,22 @@ def get_queues():
 
 
+def _add_fuzzers_to_job_items(items):
+  """Return a dic of job items along with associated fuzzers."""
```
s/dic/dict/.

iamarshsingh replied:
done

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -67,9 +83,11 @@ def get_results(this):
       page=page, page_size=PAGE_SIZE, projection=None, more_limit=MORE_LIMIT)
 
+  result_items = _add_fuzzers_to_job_items(items)
+
```
no need for this newline.

iamarshsingh replied:
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -123,6 +143,8 @@ def post(self):
           400)
 
+    fuzzers = self.request.get('fuzzers', []).split(',')
+
```
same here. no need for the newline.

iamarshsingh replied:
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -170,6 +192,8 @@ def post(self):
     job.put()
 
+    fuzzer_selection.update_mappings_for_job(job, fuzzers)
+
```
no need for newline.

iamarshsingh replied:
done 👍

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -53,6 +53,22 @@ def get_queues():
 
+def _add_fuzzers_to_job_items(items):
+  """Return a dic of job items along with associated fuzzers."""
+  result_items = []
```
Would look a bit nicer implemented like this and avoid the explicit loop:

```python
def _job_to_dict(job):
  result = job.to_dict()
  ...
  return result
```

then...

```python
...
'items': [_job_to_dict(item) for item in items]
...
```

in `get_results`.

iamarshsingh replied:
Should I add the fuzzer addition part in _job_to_dict itself?

iamarshsingh replied:
done

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -98,6 +117,7 @@ def get(self):
             'templates': templates,
             'fieldValues': {
                 'csrf_token': form.generate_csrf_token(),
+                'fuzzers': fuzzers,
```
seems like there's no need for the fuzzers variable. let's just use data_handler.get_all_fuzzer_names_including_children() inline here.

iamarshsingh replied:
done 👍

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
 
+def update_mappings_for_job(job, mappings):
+  """Clear existing mappings for a job, and replace them."""
+
```
remove newline.

iamarshsingh replied:
done 👍

oliverchang commented on src/python/tests/appengine/handlers/jobs_test.py:
```diff
@@ -81,6 +81,21 @@ def test_post(self):
     resp = self.app.post_json('/', {'page': 4})
     self.assertListEqual([], [item['id'] for item in resp.json['items']])
 
+  def test_fuzzers_result(self):
```
Also add a test for job update. You can mock out `update_mappings_for_job` if that makes things easier and just assert that it was called with the right values.

iamarshsingh replied:
yes, I have added a particular test for this and I will try to make another PR to cover the non-unit tested handlers to get more practice on working with mocks. pretty cool stuff!

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+  """Clear existing mappings for a job, and replace them."""
+
+  old_fuzzers = data_types.Fuzzer.query(data_types.Fuzzer.jobs == job.name)
+  old_mappings = {}
```
rename this `existing_fuzzers`

iamarshsingh replied:
done

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+def update_mappings_for_job(job, mappings):
+  """Clear existing mappings for a job, and replace them."""
+
+  old_fuzzers = data_types.Fuzzer.query(data_types.Fuzzer.jobs == job.name)
```
rename this existing_fuzzers_query.

iamarshsingh replied:
done

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+
+    fuzzer = data_types.Fuzzer.query(
+        data_types.Fuzzer.name == fuzzer_name).get()
+
```
remove newline. try to be conservative here and reduce vertical whitespace where we can.

iamarshsingh replied:
done 👍

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+
+  for fuzzer in old_mappings.values():
+    fuzzer.jobs.remove(job.name)
+    fuzzer.put()
```
Let's do a single ndb.put_multi(fuzzers) at the end somewhere rather than all these individual put() calls.

iamarshsingh replied:
done

iamarshsingh commented:
/gcbrun

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+  old_fuzzers = data_types.Fuzzer.query(data_types.Fuzzer.jobs == job.name)
+  old_mappings = {}
+  for fuzzer in old_fuzzers:
+    old_mappings[fuzzer.name] = fuzzer
```
We can do this:

```python
existing_fuzzers = {
  fuzzer.name: fuzzer for fuzzer in existing_fuzzers_query
}
```

to have a more declarative style that's easier to read.

iamarshsingh replied:
done, just added everything in a single line to avoid lint errors.

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +62,37 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+    fuzzer.put()
+    update_mappings_for_fuzzer(fuzzer)
+
+  for fuzzer in old_mappings.values():
```
Add a comment here to say something along the lines of "remaining values of `existing_fuzzers` are no longer mapped. remove them."

iamarshsingh replied:
done

inferno-chromium commented:
Ping me for a review once you have addressed Oliver's comments.

iamarshsingh commented:
/gcbrun

oliverchang commented on src/python/tests/appengine/handlers/jobs_test.py:
```diff
@@ -13,13 +13,15 @@
 # limitations under the License.
 """Tests for jobs."""
 import collections
+import mock
```
Use `from unittest import mock` instead to use the one from standard libs.

iamarshsingh replied:
done 👍

oliverchang commented on src/python/fuzzing/fuzzer_selection.py:
```diff
@@ -62,6 +63,36 @@ def update_mappings_for_fuzzer(fuzzer, mappings=None):
+      continue
+
+    fuzzer.jobs.append(job.name)
+    fuzzer.put()
```
Let's use a single put_multi that puts both these and the ones on line 93.

iamarshsingh replied:
done 👍

oliverchang commented on src/python/tests/core/fuzzing/fuzzer_selection_test.py:
```diff
@@ -123,6 +129,92 @@ def test_mapping_removed(self):
+    self.assertIn('fuzzer_1', mappings)
+    self.assertIn('fuzzer_2', mappings)
+
+  def test_mapping_subsituted(self):
```
nit: fix spelling of subsituted. Should be "substituted".

iamarshsingh replied:
done 👍

iamarshsingh commented:
/gcbrun

oliverchang commented on src/python/tests/core/fuzzing/fuzzer_selection_test.py:
```diff
@@ -182,7 +182,7 @@ def test_mapping_added(self):
 
-  def test_mapping_subsituted(self):
+  def test_mapping_substituted(self):
     """Ensure that mappings are subsituted properly."""
```
nit: fix the docstring too.

iamarshsingh replied:
done 👍

iamarshsingh commented:
/gcbrun

oliverchang commented on src/appengine/private/components/jobs/multi-listbox.html:
```diff
@@ -0,0 +1,81 @@
+      validate: function() {
+        const isValid = !this.required || !!(this.value && this.value.length > 0);
+        this.invalid = !isValid;
+        console.log('invalid', this.invalid);
```
remove this.

iamarshsingh replied:
done

iamarshsingh commented:
/gcbrun

oliverchang commented on src/appengine/handlers/jobs.py:
```diff
@@ -97,11 +107,18 @@ def get(self):
+                'csrf_token':
+                    form.generate_csrf_token(),
+                'fuzzers':
+                    data_handler.get_all_fuzzer_names_including_children(),
```
This should be something like:
```python
fuzzers = [
        fuzzer.name for fuzzer in data_types.Fuzzer.query(projection=['name'])
]
```
get_all_fuzzer_names_including_children isn't the right one to use, as it includes child fuzzers. child fuzzer -> job associations are set to the parents, and not modifiable .

iamarshsingh replied:
done 👍

iamarshsingh commented:
/gcbrun

oliverchang commented:
Nice, everything seems to work now :) Merging!
